Advanced Settings to React/EUI

jen-huang opened the pull request:
Fixes #17045
Relates to #14791

This PR refactors the **Management > Advanced Settings** page to use React and EUI.

## Table -> Form
Advanced Settings is now displayed as a page-long form, instead of a table. This was done after seeing the direction of the designs for Kibana v7.

<img width="1129" alt="screen shot 2018-03-29 at 2 37 41 pm" src="https://user-images.githubusercontent.com/1965714/38114858-d27a0c62-335e-11e8-88c7-fead2216b2f0.png">

## Categories
The initial page is extremely long, separated by **Categories**. The search bar can be used to filter settings like before, but there is now also a Category filter - this should hopefully reduce the user's need to scroll forever to find a setting. This paves work for v7 as well, though v7 will likely use submenu navigation for categories rather than filtering. Categorization is done by adding a new property `category` to each setting. Settings without a `category` are treated as `general` for purposes of this display.

Categories are listed alphabetically, except for General, which floats to the top.

![mar-29-2018 14-43-30](https://user-images.githubusercontent.com/1965714/38115068-9b963a12-335f-11e8-9094-5faaee7e548d.gif)

## Editing
Since all of the settings are now displayed as inline forms, there is no need to click an Edit button. Just start changing the value, and Save/Cancel actions will display next to the field. All field types support **Escape** to Cancel and most fields support **Enter** to Save (textarea and image fields do not).

![mar-29-2018 14-52-32](https://user-images.githubusercontent.com/1965714/38115422-dee52002-3360-11e8-9594-e5c0f797741b.gif)

JSON settings are validated. Settings can be cleared (reset back to default value) by clicking Reset below the field.

![mar-29-2018 15-04-11](https://user-images.githubusercontent.com/1965714/38115865-92a9351e-3362-11e8-8780-1ba85adc23d1.gif)

If the setting is an image and there isn't an image set, the file picker will display. **If an image exists,** it will be displayed with a Change link option below it. Image fields validate on max size, if there is one.

![mar-29-2018 15-13-33](https://user-images.githubusercontent.com/1965714/38116267-330fad20-3364-11e8-9214-874f6a8043c3.gif)

jen-huang commented:
Here is how the current settings are categorized:

```
General:
    context:defaultSize
    context:step
    context:tieBreakerFields
    courier:customRequestPreference
    courier:setRequestPreference
    csv:quoteValues
    csv:separator
    dateFormat
    dateFormat:dow
    dateFormat:scaled
    dateFormat:tz
    defaultIndex
    fields:popularLimit
    filterEditor:suggestValues
    filters:pinnedByDefault
    format:bytes:defaultPattern
    format:currency:defaultPattern
    format:defaultTypeMap
    format:number:defaultLocale
    format:number:defaultPattern
    format:percent:defaultPattern
    histogram:barTarget
    histogram:maxBars
    history:limit
    indexPattern:fieldMapping:lookBack
    indexPattern:placeholder
    indexPatterns:warnAboutUnsupportedTimePatterns
    metaFields
    metrics:max_buckets
    query:allowLeadingWildcards
    query:queryString:options
    savedObjects:listingLimit
    savedObjects:perPage
    search:queryLanguage
    shortDots:enable
    sort:options
    state:storeInSessionStorage
    telemetry:optIn
    timepicker:quickRanges
    timepicker:refreshIntervalDefaults
    timepicker:timeDefaults
    truncate:maxHeight
    xPack:defaultAdminEmail
Visualization:
  courier:ignoreFilterIfFieldNotInIndex
  visualization:colorMapping
  visualization:dimmingOpacity
  visualization:loadingDelay
  visualization:regionmap:showWarnings
  visualization:tileMap:WMSdefaults
  visualization:tileMap:maxPrecision
  visualize:enableLabs
Discover:
  courier:maxSegmentCount
  defaultColumns
  discover:aggs:terms:size
  discover:sampleSize
  discover:sort:defaultOrder
  doc_table:highlight
Dashboard:
  dashboard:defaultDarkTheme
  xpackDashboardMode:roles
Notifications:
  notifications:banner
  notifications:lifetime:banner
  notifications:lifetime:error
  notifications:lifetime:info
  notifications:lifetime:warning
Timelion:
  timelion:default_columns
  timelion:default_rows
  timelion:es.default_index
  timelion:es.timefield
  timelion:graphite.url
  timelion:max_buckets
  timelion:min_interval
  timelion:quandl.key
  timelion:showTutorial
  timelion:target_buckets
Reporting:
  xpackReporting:customPdfLogo
```

snide commented:
Oh man, this looks fantastic

alexfrancoeur commented:
Woah! nice @jen-huang I'll pull down this week and play around with it. From your gif's, this looks fantastic

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1259/)

skearns64 commented:
This is great!  :thumbsup: :thumbsup:

timroes commented:
This is awesome! Thanks a lot :heart: 

Just a few points I would like to discuss and hear your opinion on:

1) I haven't had the time to test it out yet, so I just simply ask: do we have the possibility to link to individual settings, that the page is already filtered for them?
2) I know this would cause some more work, but could we maybe besides the current id and description also add a "display name" (i.e. title, short name) for each setting? I would imagine, that this title would be shown above the element instead of the current id, and the id will just be visible is parentheses behind it, or via a tooltip. Some examples: `CSV Separator` (for `csv:separator`), `Timezone` (for `dateFormat:tz`), `Start of week` (for `dateFormat:dow`). That would make the settings even more user friendly in my opinion. Should be rather easy implementation wise, the hard step would be coming up with a name for all settings :-)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1269/)

chrisronline commented:
Pinging @elastic/kibana-visualizations @elastic/kibana-sharing @elastic/kibana-discovery to verify [this separation](https://github.com/elastic/kibana/pull/17465#issuecomment-377391256) makes sense to you

chrisronline commented:
This looks AWESOME!

I just a couple questions on the design:

There is quite a bit of dead space on the right. Maybe we can horizontally stack the areas if there is real estate?

<img width="1378" alt="screen shot 2018-03-30 at 9 46 56 am" src="https://user-images.githubusercontent.com/56682/38139886-611230ee-33ff-11e8-8a38-a10e8a75e2ef.png">

It feels like the title should stand out more than it does. It's hard to separate the text visually when they are the same size/weight:
<img width="446" alt="screen shot 2018-03-30 at 9 47 53 am" src="https://user-images.githubusercontent.com/56682/38139905-7d5f81d4-33ff-11e8-93c0-7b220c5a3f26.png">

nreese commented:
This looks really nice.

One comment about the categories is that some settings span more than one category. It would make sense to have the category be an array so settings can be part of multiple categories.

For example, `courier:ignoreFilterIfFieldNotInIndex` effects both visualize and dashboard.

Bargs commented:
The `context` options probably belong under Discover. I'd probably put all of the `courier` stuff under general, or maybe even under it's own category, like `search` or something. `courier:ignoreFilterIfFieldNotInIndex` for example applies everywhere, not just under Visualize, the same is true of most of those courier settings. Categorizing some of these settings is a tough call.

Looks like a really nice improvement, that screen was getting pretty crazy.

thomasneirynck commented:
wrt. categorization. imho this remains OK for visualizations.

snide commented:
Hey @jen-huang just some very minor designy suggestions. Feel free to take um or leave um.

To help give the form some room, I'd go ahead and use a max-width panel with a background to match the styling we have in EUI. Should help keep it from being jammed next to the left side.

![image](https://user-images.githubusercontent.com/324519/38205299-fa9c30ae-365a-11e8-9e2f-8d74d00cb670.png)

On this dropdown, I'd probably keep the other sections around for selection, even if only one of them can be selected. Right now you have to clear a selection before you can select another one.

![image](https://user-images.githubusercontent.com/324519/38205378-3d18ac5a-365b-11e8-9eb2-cfc5be71fb74.png)

You might want to set your flexItems to grow={false}, then set the cancel button to be an empty button. Should make them a little less in your face and size appropriately.

![image](https://user-images.githubusercontent.com/324519/38205522-cfe270f2-365b-11e8-81ca-344c42a91e83.png)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1310/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1320/)

jen-huang commented:
Thanks for the feedback, all! PR has been updated with the following changes:

- `search` has been added as a category, containing all the `courier` settings. Moved `context` settings under `discover`. (@Bargs)
- Changed `category` property to be an array, so that in the future when settings are broken into multiple category pages, we can display the same setting under each category it belongs to. Right now everything is still on one page and it doesn't make sense to display a setting multiple times, so each setting just has one entry in its `category` array. (@nreese)
- Each setting now has a `name` property meant for displaying a user-friendly version of the setting key. [List has been updated containing the names](https://github.com/elastic/kibana/pull/17465#issuecomment-377391256). (open to changes!) (@timroes)
- Added deep-link route to a specific setting: `#/management/kibana/settings/[key]`. For example, `#/management/kibana/settings/notifications:banner` will display the page filtered down to the `notifications:banner` setting. (@timroes)
- Updated UI with @snide's great suggestions. This improves the empty space on the right. #17467 has also been merged which fixes bolding of the field labels. (@chrisronline)

UI changes with new friendly names:

<img width="834" alt="screen shot 2018-04-02 at 1 36 56 pm" src="https://user-images.githubusercontent.com/1965714/38214934-f4d97cce-367a-11e8-941c-0a611e27c4de.png">

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1329/)

chrisronline commented:
This is awesome!

When I filter down to a single category, then I reopen the selection list, I only see the one I previously selected which doesn't let me select multiple without unselecting, then reselecting once the entire list is repopulated.

<img width="638" alt="screen shot 2018-04-03 at 11 07 08 am" src="https://user-images.githubusercontent.com/56682/38257893-2e3de59a-372f-11e8-8bba-545ec8efa544.png">

Also, maybe we should be alternating background colors, or adding `<hr>` tags to separate rows because it's not immediately clear to me in this screenshot what I changed or am saving:

<img width="985" alt="screen shot 2018-04-03 at 11 08 40 am" src="https://user-images.githubusercontent.com/56682/38258012-731b9a68-372f-11e8-8964-f303ba863a7a.png">

jen-huang commented:
@chrisronline 
Thanks for catching the filter dropdown bug - fixed now!

I played around with some options for differentiating between field rows. It's really only an issue for the boolean settings, since there is a large gap between the `EuiSwitch` toggle and the Save button. The other types of settings are fine since their input fields take up the full 400px. Alternating colors / `hr` seemed too jarring when I tried, so instead I decreased the gutter spacing next to Save, and increased spacing between each setting:

<img width="704" alt="screen shot 2018-04-03 at 2 11 55 pm" src="https://user-images.githubusercontent.com/1965714/38276179-0744d8a8-3749-11e8-92c3-53e8cf4c85f8.png">

Hopefully this makes the distinction a little clearer. It's not perfect, but since the Save/Cancel buttons pop up immediately when editing, I'm not too concerned about the user not knowing what changed.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1370/)

nreese commented:
I am concerned about the confusion caused by having 2 different search bar implementations. Kibana already has kuery for the filter bar. Kuery has a different syntax than EuiSearchBar. For example EuiSearchBar uses `-` to negate terms while kuery uses `not`. Is there a way to reconcile the search grammers between the two?

alexfrancoeur commented:
@jen-huang I wanted to take a look to provide some feedback but am running into some issues pulling it down on a fresh version of kibana. Any reason I'd be getting these errors?

```
From git://github.com/jen-huang/kibana
 * branch                  eui/advanced-settings -> FETCH_HEAD
Auto-merging yarn.lock
CONFLICT (content): Merge conflict in yarn.lock
Auto-merging src/core_plugins/timelion/index.js
Removing src/core_plugins/kibana/public/management/sections/settings/lib/get_editor_type.js
Auto-merging src/core_plugins/kibana/public/management/sections/settings/lib/__tests__/to_editable_config.test.js
Auto-merging src/core_plugins/kibana/public/management/sections/settings/lib/__tests__/get_val_type.test.js
Removing src/core_plugins/kibana/public/management/sections/settings/lib/__tests__/get_editor_type.js
Auto-merging src/core_plugins/kibana/public/management/sections/settings/lib/__tests__/get_aria_name.test.js
Removing src/core_plugins/kibana/public/management/sections/settings/advanced_row.js
Removing src/core_plugins/kibana/public/management/sections/settings/advanced_row.html
Auto-merging package.json
CONFLICT (content): Merge conflict in package.json
Automatic merge failed; fix conflicts and then commit the result.
```

jen-huang commented:
@alexfrancoeur I updated branch with master, that should resolve the dependency conflicts. You'll also need the corresponding x-pack-kibana branch if running x-pack: https://github.com/elastic/x-pack-kibana/pull/5089.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1625/)

alexfrancoeur commented:
Thanks @jen-huang! This PR is amazing and already so much better than the experience we have today. I did have a bit of feedback though, let me know what you think. Below you'll find some comments on my experience with the new look and feel. Feel free to use them as you will, hope this helps!

We seem to be wasting a fair bit of whitespace here. It looks a little off being aligned and squished to the left. Do you think it's possible we could make it a bit more responsive? Might be good to quickly get designs input here from @cchaos or @snide 
<img width="1282" alt="screen shot 2018-04-10 at 2 56 43 pm" src="https://user-images.githubusercontent.com/25181823/38577701-b1c042d2-3ccf-11e8-9ad1-9da80e37acf2.png">

I noticed that there was no "all" option to revert the selection in the dropdown. It would be nice to be able to select an option from the drop down for this rather than remove the search from the search bar. 

<img width="576" alt="screen shot 2018-04-10 at 2 57 18 pm" src="https://user-images.githubusercontent.com/25181823/38577714-b7ce4926-3ccf-11e8-9fe6-da0c657f67f8.png">

If there is just a switch, should we add some text? Such as "enable / disable"? I'm not sure if that's part of the component but it's always seemed odd to me that we just let them float there.

<img width="438" alt="screen shot 2018-04-10 at 3 55 28 pm" src="https://user-images.githubusercontent.com/25181823/38580338-ae54d290-3cd7-11e8-9faa-a17678f33510.png">

Not sure if this is an EUI component thing (it might be), but is the descriptive text of the input control always going to be below it instead of below the title?

<img width="466" alt="screen shot 2018-04-10 at 3 58 54 pm" src="https://user-images.githubusercontent.com/25181823/38580508-23eafb56-3cd8-11e8-9248-58bcc2d98d96.png">

Does it make sense to use the code editor EUI component than a text box for the JSON components?
https://elastic.github.io/eui/#/forms/code-editor. I'm guessing we just aren't differentiating now.

<img width="440" alt="screen shot 2018-04-10 at 4 05 54 pm" src="https://user-images.githubusercontent.com/25181823/38580846-13c7251e-3cd9-11e8-86f8-198c2678920e.png">

Is there a reason these are clickable? More importantly, if we have a human readable title, do we need to technical one out in the open? Maybe this becomes a tooltip or something?

<img width="195" alt="screen shot 2018-04-10 at 4 06 30 pm" src="https://user-images.githubusercontent.com/25181823/38580902-315b4182-3cd9-11e8-9ff1-acbe5ce32627.png">

Now that we have human readable titles for each setting, should we sort by them? It looks as if we're still sorting by the variable name

<img width="549" alt="screen shot 2018-04-10 at 4 08 19 pm" src="https://user-images.githubusercontent.com/25181823/38581018-8d1aae2c-3cd9-11e8-9c7b-b56899e25103.png">

jen-huang commented:
@alexfrancoeur Thanks for the feedback!

- The Save/Cancel buttons appear to the right of the fields after editing, so there will always be some whitespace needed, but I will play around with the panel width.
- Help text will always appear under the fields since we are using `EuiFormRow` here https://elastic.github.io/eui/#/forms/form-layouts.
- The readable name and setting key are both part of the field label, so clicking on either gives focus to the corresponding input field.
- I kept the display of the setting key (e.g. `dateFormat:scaled`) because [our documentation](https://www.elastic.co/guide/en/kibana/current/advanced-options.html), and other community resources ([example](https://discuss.elastic.co/t/how-can-i-change-the-x-axis-labels-for-a-date-format-in-kibana/564/2)), direct the user to look for and change a setting based its key. Let me know what you think. It may be a nonissue if the readable names are clear enough for what key they are representing (and we can move the key to tooltip for extra clarification).
- On a related note, I kept the sorting by key mostly for the General category so that related settings are next to each other. E.g. the `dateFormat` settings are grouped together when sorting by key, but not by readable name.

cchaos commented:
I think this is a great example of needing a more horizontal layout for long forms and form rows. A possibility would be to separate the humanized label and help text from the row itself and put it in a column next to the field. You can also then break down the panels/pages by settings section and keep the actual settings stuff like title, search capability, callout, etc.. outside of the page itself.

<img width="1502" alt="screen shot 2018-04-16 at 12 26 38 pm" src="https://user-images.githubusercontent.com/549577/38822557-c946e82e-4171-11e8-9ab0-b53bb33008ae.png">


Just a thought, @snide?

snide commented:
@cchaos I think that layout looks pretty good. I'd worry a bit about the separated paneling, just because search might return only one result in a couple different panels so it might make sense just to use whitespace within a single panel. Otherwise that all looks composable from our current EUI suit with some flexboxes and titles.

cchaos commented:
I think the separation using panels will actually help when searching has been applied as it would be easier to tell which sections are not showing all settings. I also like the idea of adding some text per section explaining this behavior and allowing a quick clear function.

<img width="1504" alt="screen shot 2018-04-16 at 13 05 25 pm" src="https://user-images.githubusercontent.com/549577/38824319-dc709efe-4176-11e8-92b8-646ddb47d042.png">

timroes commented:
I came across #https://github.com/elastic/kibana/issues/9673

timroes commented:
I came across #9673. Is this something, that is touched enough by that PR, or if it's not you might still want to consider closing that issue, since the new management design won't include the request?

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/2250/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/2730/)

jen-huang commented:
UI updated:

- Categories separated out to different panels
- Each setting uses new `EuiDescribedFormGroup` component (human title and description to the left of field)
- Textarea fields replaced with code editor

<img width="1047" alt="screen shot 2018-05-07 at 12 41 55 pm" src="https://user-images.githubusercontent.com/1965714/39721348-213ce03c-51f4-11e8-84a4-434896b9285a.png">
